timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -0,0 +1,186 @@
+            // Need the weights to sample a value and the log probs are just because they aren't expensive to copy
+            k_store_accepted_log_probability<RealType><<<1, tpb, 0>>>(
+                num_target_mols_,
+                d_translations_.data + 3, // Offset to get the last value for the acceptance criteria
+                d_log_sum_exp_before_.data,
+                d_log_sum_exp_after_.data,
+                d_log_weights_before_.data,
+                d_log_weights_after_.data);
+            gpuErrchk(cudaPeekAtLastError());
+        }
+        // Make a copy of the coordinates
+        gpuErrchk(cudaMemcpyAsync(
+            d_intermediate_coords_.data, d_coords, d_intermediate_coords_.size(), cudaMemcpyDeviceToDevice, stream));
+
+        // Quaternions generated from normal noise, while translations and the acceptance value are uniform
+        curandErrchk(templateCurandNormal(cr_rng_, d_quaternions_.data, d_quaternions_.length, 0.0, 1.0));
+        // The last translation value is to be used to determine acceptance
+        curandErrchk(templateCurandUniform(cr_rng_, d_translations_.data, d_translations_.length));
+
+        sampler_.sample_device(num_target_mols_, num_samples, d_log_weights_before_.data, d_samples_.data, stream);
+        k_rotate_and_translate_mols<RealType><<<ceil_divide(num_samples, tpb), tpb, 0, stream>>>(
+            num_samples,
+            d_coords,
+            d_box,
+            d_samples_.data,
+            d_mol_offsets_.data,
+            d_quaternions_.data,
+            d_translations_.data,
+            d_intermediate_coords_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        mol_potential_.mol_energies_device(
+            N,
+            num_target_mols_,
+            d_intermediate_coords_.data, // Use the moved coords
+            d_params_.data,
+            d_box,
+            d_mol_energy_buffer_.data, // Don't need to zero, will be overridden
+            stream);
+
+        k_compute_log_weights_from_energies<RealType><<<mol_blocks, tpb, 0, stream>>>(
+            num_target_mols_, beta_, d_mol_energy_buffer_.data, d_log_weights_after_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        logsumexp_.sum_device(num_target_mols_, d_log_weights_after_.data, d_log_sum_exp_after_.data, stream);
+
+        k_attempt_exchange_move<RealType><<<ceil_divide(N_, tpb), tpb, 0, stream>>>(
+            N,
+            d_translations_.data + 3, // Offset to get the last value for the acceptance criteria
+            d_log_sum_exp_before_.data,
+            d_log_sum_exp_after_.data,
+            d_intermediate_coords_.data,
+            d_coords,
+            d_num_accepted_.data);
+        gpuErrchk(cudaPeekAtLastError());
+        num_attempted_++;
+    }
+}
+
+template <typename RealType>
+std::array<std::vector<double>, 2>
+BDExchangeMove<RealType>::move_host(const int N, const int num_moves, const double *h_coords, const double *h_box) {
```

Comment:
Should this match the API of the the MonteCarloMove so that it is `move(coords, box)` and instead pass in a number of proposals per call to move?

Reply:
Inclined to make it match, so we can define a `MonteCarloMove` base class in C++ that has the signature `move_device(coords, box, stream)` (or another name if there is a suggestion).

Reply:
I'd

1) Keep the API of the python layer simple, for multiple moves, there are multiple variations (multiple sequential moves, multiple-try-metropolis type of moves, etc), mixing and matching different kinds of moves (eg. HREX moves). The `moves.py` file defines some of these

2) C++ is up to you! (Not sure if we want consistency with the barostat moves, HREX is implemented purely in python right now)

Reply:
> Not sure if we want consistency with the barostat moves

Think they should be consistent, that way we can have a series of MC moves in with the context. Will need to change a few things to make it all work, but as we have more MC moves would be nice to have a singular API.